### DISCUSSION TITLE
How to shutdown server with code in linux

### DISCUSSION THREAD

**Discussion by hanaTsuk1** [static-web-server/static-web-server#315] [Q&A] [answered]
In windows, I can use `windows_service` and `cancel_recv` to shutdown server
I have read the relevant code, but in linux, I don't know how to do it
Could you please provide some assistance

**Comment by joseluisq** [comment]
Unfortunately, there is no exposed API to shout down the server gracefully on demand.
You have to send to the process a termination signal like `SIGTERM`.

**Reply by hanaTsuk1** [reply]
Thanks for your reply
Will this feature be added in the future?

**Reply by joseluisq** [reply]
I do not know but you can always suggest something or open a feature request explaining your use case.

**Reply by hanaTsuk1** [reply]
I am creating a small tool based on static web server
Here is my use case
https://github.com/yuri-app/yuri/blob/master/src-tauri/src/main.rs#L40

I am a beginner in Rust, so I am not sure if the following ideas are correct or not
Why not add a `cancel_recv` to `wait_for_signals` so that all platforms can directly control the server
https://github.com/static-web-server/static-web-server/blob/master/src/server.rs#L602

**Reply by joseluisq** [reply, marked as answer]
> Why not add a `cancel_recv` to `wait_for_signals` so that all platforms can directly control the server https://github.com/static-web-server/static-web-server/blob/master/src/server.rs#L602

Because that function was initially thought to just wait for Linux/Unix termination signals only. Later on, Windows support came along which does not have a termination signal concept and uses a channel to wait for an incoming ctrl+c. That's why both defer in implementation.
Anyway, I agree that adding support for a `cancel_recv` should be convenient for library users to shout down the server gracefully on demand.

I could propose something in the coming weeks, so stay tuned.

**Reply by hanaTsuk1** [reply]
Thanks again for your detailed reply

**Comment by joseluisq** [comment]
@hanaTsuk1 there is a PR https://github.com/static-web-server/static-web-server/pull/319 ready to be tested. Please give it a try and let me know how it goes.

**Reply by hanaTsuk1** [reply]
It works as expected

**Reply by joseluisq** [reply]
Great, it is merged now. This will be available in next upcoming release.

**Reply by joseluisq** [reply]
Released in [v2.28.0](https://github.com/static-web-server/static-web-server/releases/tag/v2.28.0)

**Reply by hanaTsuk1** [reply]
🎉🎉🎉
